src/util/misc.rs

```diff
@@ -160,17 +194,37 @@ mod message_signing {
             address: &Address,
             msg_hash: sha256d::Hash
         ) -> Result<bool, secp256k1::Error> {
+            // Mostly ported from: https://github.com/bitcoinjs/bitcoinjs-message/blob/c43430f4c03c292c719e7801e425d887cbdf7464/index.js#L181-L233
             let pubkey = self.recover_pubkey(&secp_ctx, msg_hash)?;
-            Ok(match address.address_type() {
-                Some(AddressType::P2pkh) => {
-                    *address == Address::p2pkh(&pubkey, address.network)
-                }
-                Some(AddressType::P2sh) => false,
-                Some(AddressType::P2wpkh) => false,
-                Some(AddressType::P2wsh) => false,
-                Some(AddressType::P2tr) => false,
-                None => false,
-            })
+
+            // Use .serialize because we always want compact serialization
+            let pubkey_hash = super::hash160(&pubkey.key.serialize().to_vec()).to_vec();
```

Comment:
@tcharding I'm going to go with
```
use crate::util::misc::{bech32_decode, segwit_redeem_hash, hash160, get_payload_bytes};
```
and remove the use of `super` outside of the tests. Let me know if that is sufficient.

Will look into `to_vec()` usage shortly

Reply:
Resolved `super` usage in 5ea6b6a1

Reply:
@Kixunil Do you have any specific suggestions here? I was trying to be consistent with the existing `signed_msg_hash(msg)`

```rust
/// Hash message for signature using Bitcoin's message signing format.
pub fn signed_msg_hash(msg: &str) -> sha256d::Hash {
    let mut engine = sha256d::Hash::engine();
    engine.input(BITCOIN_SIGNED_MSG_PREFIX);
    let msg_len = encode::VarInt(msg.len() as u64);
    msg_len.consensus_encode(&mut engine).unwrap();
    engine.input(msg.as_bytes());
    sha256d::Hash::from_engine(engine)
}
```

Reply:
Will need to look at it better to write the exact code (hopefully tomorrow) but generally, hasher can be considered a writer and data can be considered reader.

The latter `to_vec` shouldn't be needed because you only need slice later and borrowing rules allow it. (Unless I'm missing something very subtle - LMK if you get a compile error.)

Reply:
Actually, [`serialize()` produces `Vec<u8>`](https://docs.rs/bitcoin/0.27.1/bitcoin/util/ecdsa/struct.PublicKey.html#impl-Serialize) this is not great but proves `to_vec` is not needed because it's essentially a `clone()`.

Reply:
That link is `PublicKey.serialize` which is easily confused with `PublicKey.key.serialize()` (and PublicKey.key.serialize_uncompressed()). `PublickKey.key`is a `secpk256k1::PublicKey`. Removing the `to_vec()` in `&pubkey.key.serialize().to_vec()` results in

```
error[E0308]: mismatched types
   --> src/util/misc.rs:200:39
    |
200 |             let pubkey_hash = hash160(&pubkey.key.serialize()).to_vec();
    |                                       ^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::vec::Vec`, found array `[u8; 33]`
    |
    = note: expected reference `&std::vec::Vec<u8>`
               found reference `&[u8; 33]`

error: aborting due to previous error
```

Reply:
Implemented it here: https://github.com/Kixunil/rust-bitcoin/commit/858c83f684d8a4f38654dd7686660fa2e6ee1cc1 (I can't make a PR to your branch, no clue why).

> is easily confused with `PublicKey.key.serialize()`

Oh, that's even better as there's no `Vec`, so no more work needed to remove the allocation!

Reply:
Nice! Thanks. This is great to see.

> I can't make a PR to your branch, no clue why
Probably a permission issue on my fork? Anyway I manage to create a PR and merge it (https://github.com/Shatnerz/rust-bitcoin/pull/1)